File: lee/src/main/java/org/example/lee/题目/图/b_岛屿数量.go
package 图

func numIslands(grid [][]byte) int {
	sum := 0
	for i := 0; i < len(grid); i++ {
		for j := 0; j < len(grid[0]); j++ {
			//一个个遍历就好
			if grid[i][j] == '1' {
				tool(grid, i, j)
				sum++
			}
		}
	}
	return sum
}
func tool(grid [][]byte, x, y int) {
	if x < 0 || x >= len(grid) || y < 0 || y >= len(grid[0]) || grid[x][y] != '1' {
		return
	}
	grid[x][y] = '0'
	tool(grid, x-1, y)
	tool(grid, x+1, y)
	tool(grid, x, y-1)
	tool(grid, x, y+1)
}

File: lee/src/main/java/org/example/lee/题目/图/b_岛屿数量.java
package org.example.lee.题目.图;

public class b_岛屿数量 {

	/**
	 * num islands 给你一个由 '1'（陆地）和 '0'（水）组成的的二维网格，请你计算网格中岛屿的数量。
	 * 岛屿总是被水包围，并且每座岛屿只能由水平方向和/或竖直方向上相邻的陆地连接形成。 此外，你可以假设该网格的四条边均被水包围。
	 *
	 * @param grid grid
	 * @return int
	 */
	public int numIslands(char[][] grid) {
		int len = grid.length, key = 0;
		for (int i = 0; i < len; i++) {
			for (int j = 0; j < grid[0].length; j++) {
				if (grid[i][j] == '1') {
					dfs(grid, i, j);
					key++;
				}
			}
		}
		return key;
	}

	public void dfs(char[][] grid, int i, int j) {
		if (isArea(grid, i, j) && grid[i][j] == '1') {
			grid[i][j] = '2';
			dfs(grid, i, j - 1);
			dfs(grid, i, j + 1);
			dfs(grid, i + 1, j);
			dfs(grid, i - 1, j);
		}
	}

	public boolean isArea(char[][] grid, int i, int j) {
		return i >= 0 && j < grid[0].length && j >= 0 && i < grid.length;
	}
}
